Pretty print journal entries

alichtman opened the pull request:
Changes appearance of all jrnl viewing commands, such as `$ jrnl --short` and
`$ jrnl -n {NUM}`.

Fix #508, #717

alichtman commented:
This feature has some slight bugs that I'm still working out.

alichtman commented:
The last failing test case is exporting with custom templates. This is also failing for me on master, so I don't think it's my PR causing this.

```
  Scenario: Exporting using custom templates  # features/exporting.feature:23
    Given we use the config "basic.yaml"      # features/steps/core.py:68 0.003s
    Given we load template "sample.template"  # features/steps/core.py:102 0.001s
    When we run "jrnl --export sample"        # features/steps/core.py:92
NotImplementedError
   "-" * len(entry.title)
'Constant' not supported
NotImplementedError
   "-" * len(entry.title)
    When we run "jrnl --export sample"        # features/steps/core.py:92 0.022s
    Then the output should be                 # features/steps/core.py:171 0.000s
      """
      My first entry.
      ---------------

      Everything is alright

      Life is good.
      -------------

      But I'm better.
      """
      Assertion Failed: ['---------------', 'None']
      Captured stdout:
      My first entry.
      None

      Everything is alright

      Life is good.
      None

      But I'm better.

```

peter-schmidbauer commented:
I'm not the biggest fan of the color choices to be honest. Why not let the user define the colors (separately for date, title and body) in the config?

alichtman commented:
> Why not let the user define the colors (separately for date, title and body) in the config?

Good call. Do you think it's better to have them set colors like `red, blue, green, ...` or ANSI escapes?

peter-schmidbauer commented:
Definitely colors. Not sure what the ideal way to map them to escapes is though. We should probably look into moving away from using the raw escapes all together and towards a package like [colorama](https://pypi.org/project/colorama/).

alichtman commented on features/steps/core.py:
```diff
@@ -5,8 +5,10 @@
+from ansiwrap import strip_color
 from collections import defaultdict
 import os
+import re
```
This should be removed.

alichtman commented:
I should add two tests:

1. Identifying invalid color configs
2. Upgrading the config from no colors -> colors

alichtman commented on jrnl/util.py:
```diff
@@ -162,6 +162,21 @@ def scope_config(config, journal_name):
+    all_valid_colors = True
+    for key, color in config["colors"].items():
+        if not getattr(colorama.Fore, color.upper(), None):
+            # TODO: Not sure whether both of these should stay or not.
```
This should be addressed before merging.

peter-schmidbauer replied:
Probably just the print(), but use print(..., file=sys.stderr) to keep it in line with the other jrnl output.

alichtman commented:
@wren I believe this is ready for review.

wren commented:
@alichtman Great! I'll review this weekend.

One thing from first glance, though: Is there an option to disable color? I know a lot of people pipe output to various scripts, and the escape sequences can complicate that.

alichtman commented:
Yep, I can add that.

peter-schmidbauer commented on jrnl/util.py:
```diff
@@ -180,9 +198,20 @@ def get_text_from_editor(config, template=""):
+    if upper_color == "NONE":
+        return string
+
+    color_escape = getattr(colorama.Fore, upper_color, colorama.Fore.WHITE)
```
I think the default should just be no color at all. This way we keep backwards compatibility and we don't confuse users who have set their terminal background color to white.

alichtman replied:
White is the default for an invalid color, but I can change the color defaults to none when the config is created.

peter-schmidbauer replied:
Ah, I just realized somebody with an old config would just get the new default values as well.
Yea that would probably be better.

alichtman replied:
Done.

peter-schmidbauer commented on jrnl/install.py:
```diff
@@ -57,6 +58,10 @@ def module_exists(module_name):
     'indent_character': '|',
+    'colors': {
+        'date': 'red',
+        'title': 'blue',
```
I'd add one for the body and for tags too, because why not? ;)

alichtman commented:
I've found one bug in the config updating algorithm that will need to be addressed. It only looks at top-level keys, and not nested keys.

https://stackoverflow.com/a/27266178

A solution like this will be needed if the subkeys of `colors` will ever be updated...

I'll open a separate issue.

alichtman commented:
<img width="642" alt="image" src="https://user-images.githubusercontent.com/20600565/68326525-e2b38b80-00cb-11ea-9aa4-61e649197fa5.png">

So here's an interesting bug...

I understand why the body drops all colorama formatting after the tag (due to the `colorama.Style.RESET_ALL` in `colorize()`), but I can't figure out why the title doesn't have the same problem.

peter-schmidbauer commented:
By the way, I think stripping the color from all outputs during tests is not ideal. I'd instead have all previous tests use no color, and add a new configuration with colors and some more specific tests. It's a new feature and deserves its own tests imho.

alichtman commented:
I tried for a few hours to figure out how to get behave to work with ANSI escapes and made no progress. I agree with you, but I don't think it's possible.

alichtman commented on jrnl/util.py:
```diff
@@ -180,9 +196,58 @@ def get_text_from_editor(config, template=""):
+    """
+    config = entry.journal.config
+    if config['highlight']:  # highlight tags
+        if entry.journal.search_tags:
```
The code duplication here bothers me but I don't know how to clean it any more at the moment.

alichtman replied:
Figured out how to condense this, but caused a regression in the test for unreadable dates that I'm working on fixing at the moment.

alichtman commented:
<img width="641" alt="image" src="https://user-images.githubusercontent.com/20600565/68336543-eac8f680-00de-11ea-890c-35264b34995d.png">

Looks functional to me.

peter-schmidbauer commented on jrnl/util.py:
```diff
@@ -162,6 +162,22 @@ def scope_config(config, journal_name):
+    for key, color in config["colors"].items():
+        upper_color = color.upper()
+        if upper_color == "NONE":
+            pass
```
This should be "continue"

alichtman replied:
Good catch.

alichtman commented:
That last commit breaks one regression test, and I'm not sure why.

`features/regression.feature:51  Journals with unreadable dates should still be viewable`

Doesn't happen when the commit isn't applied, but I can't find anything in that commit which should cause it to break...

alichtman commented:
Found the issue. This last commit clobbers `\n` characters in the bodies of journal entries.

<img width="600" alt="image" src="https://user-images.githubusercontent.com/20600565/68353021-92a6ea00-0108-11ea-8515-1e7811484b4e.png">

alichtman commented:
This is all ready for review. No bugs that I'm aware of.

peter-schmidbauer commented:
I was unsatisfied with not being able to test it so I did a little digging... Behave does seem to escape the backslashes automatically kind of like r-strings (e.g. `print(r"\n")`). 
It seems like it's easy to "unescape" them again by doing `text.encode("utf-8").decode("unicode_escape")`:

![grafik](https://user-images.githubusercontent.com/13824347/68417642-13142c00-0197-11ea-8535-3fd1a19041b3.png)

However that doesn't seem to play nice with unicode characters...

![grafik](https://user-images.githubusercontent.com/13824347/68417708-3212be00-0197-11ea-92a2-22ec76c78d3c.png)

However, the codecs package provides a better way that works with unicode:

![grafik](https://user-images.githubusercontent.com/13824347/68417821-69816a80-0197-11ea-9c75-891c201fa400.png)

With this you can transform the text from behave and then compare it to the stdout capture without stripping any colors.

Thanks to [this answer](https://stackoverflow.com/a/57192592/5266392) on Stackoverflow!

alichtman commented:
That's a nice trick! I got stuck on that problem... I'll take a look at revamping the tests.

peter-schmidbauer commented:
It's a bit cryptic, so make sure to link to the SO post in a comment. However, in my tests it seemed to work fine. 

Slightly off topic, I just noticed a potential bug:

![grafik](https://user-images.githubusercontent.com/13824347/68426269-0a782180-01a8-11ea-8b1d-aa39e4455e67.png)

The tags in the title seem to have a slightly different color (config for tags is yellow) for some reason and they also seem to add a space behind them.

alichtman commented:
Ugh. This space bug has been haunting me... I thought I had it all fixed in that last set of commits. Lemme see if I can repro it.

alichtman commented:
This one is tricky to fix. The way I have this set up makes it difficult to distinguish between these two cases: `@tag. Text` and `@tag - text`. This branch currently breaks a regression test due to this.

peter-schmidbauer commented:
How about `re.split(r'(?<=\s)([{@}][-+*#/\w]+)', text)`?

peter-schmidbauer commented:
I'm not sure where you need that re.split, but `re.split(r'([@][-+*#/\w]+)', text)` does what you expected.

alichtman commented:
> I'm not sure where you need that `re.split`

It's broken out from `Entry.tag_regex()`, which is used in `highlight_tags_with_background_color()`.

```python
text_fragments = re.split(entry.tag_regex(config['tagsymbols']), text)
```

The bug fix will go in either this method, `colorized_text_generator`, or `colorize`.

I'm going to have to think about how this should be working some more. I've spent way too much time on this feature over the past few days and need to take care of some other stuff. If you see a way to solve the issue that causes the tests that I wrote for this feature to fail, definitely lmk. I've got nothing at the moment.

alichtman commented:
Sweet, the newline bug is fixed. Now the only bug left to fix is the missing space after a tag that is not followed by punctuation. See the first line of the body for an example.

<img width="321" alt="image" src="https://user-images.githubusercontent.com/20600565/68522053-6a092680-02a7-11ea-8ed6-1a4a3b24d677.png">

alichtman commented:
Decided to bold all tags and titles. Works great in PyCharm, doesn't work in my terminal.

<img width="535" alt="image" src="https://user-images.githubusercontent.com/20600565/68523513-9da17c00-02ba-11ea-9d23-3c32a48d9f5d.png">


<img width="439" alt="image" src="https://user-images.githubusercontent.com/20600565/68523515-a2fec680-02ba-11ea-89e1-caacd38e93b4.png">

This is a problem for another day, though.

wren commented:
Hey, we just merged a bunch of stuff into master. This just needs a quick rebase on the new master, but then should be ready to go. Thanks, and sorry!

alichtman commented:
Disclaimer: I did that merge in the GitHub UI. I'm pretty sure it's done correctly but we shall see...

alichtman commented:
Resolve merge conflict in `poetry.lock`

alichtman commented:
@wren Any update on this?

wren commented:
@alichtman I have a full jrnl work day scheduled for Saturday, so I'll be able to go fully through this then. Lots a big one, so I want to make sure to put it through its paces.

pspeter commented on jrnl/util.py:
```diff
@@ -4,21 +4,30 @@
-ERROR_COLOR = "\033[31m"
-RESET_COLOR = "\033[0m"
+PY3 = sys.version_info[0] == 3
+PY2 = sys.version_info[0] == 2
```
Don't add all these constants back in, they were purposefully removed previously.

pspeter replied:
I also meant the ones below this comment (only excluding the colors)

alichtman replied:
Whoops. Fixed.

pspeter commented on jrnl/Entry.py:
```diff
@@ -49,8 +49,8 @@ def tags(self):
-        pattern = fr'(?u)(?:^|\s)([{tagsymbols}][-+*#/\w]+)'
-        return re.compile(pattern)
+        pattern = r'(?<!\S)([{tags}][-+*#/\w]+)'.format(tags=tagsymbols)
+        return re.compile(pattern, re.UNICODE)
```
Same with some of these changes. re.UNICODE is not needed and you reverted the fstring.

alichtman replied:
I was under the impression that `(?u)` and `re.UNICODE` were the same, so that was intentional. Otherwise, these were merging errors. Sorry about that.

pspeter replied:
That's true actually, they are the same and you can safely remove both.

alichtman replied:
Initially, it was done in this commit: 95b23d433c124de685a1bd9891fdff8f1edb308a

Fixed all the merge errors. Sorry about that.

pspeter commented on features/steps/core.py:
```diff
@@ -205,7 +217,9 @@ def check_output_time_inline(context, text):
-    out = context.stdout_capture.getvalue()
+    out = process_ANSI_escapes(context.stdout_capture.getvalue())
+    if isinstance(out, bytes):
+        out = out.decode('utf-8')
```
Same here

wren commented:
This is looking good so far. I didn't quite get this one finished, but I'll wrap up on Monday night. Thanks for your patience!
